src/stirling/source_connectors/socket_tracer/bcc_bpf/socket_trace.c

```diff
@@ -151,6 +162,29 @@ static __inline void set_conn_as_ssl(uint32_t tgid, int32_t fd) {
   conn_info->ssl = true;
 }
 
+// Writes the syscall fd to a BPF map to be later consumed by the tls tracing probes in
+// addition to ensuring that set_conn_as_ssl is called. The majority of probes should call
+// this on function entry, however, there are certain probes which are not exlusively used
+// for networking and therefore must defer this fd propagation to its ret probe (where it
+// can be validated it is socket related first via sock_event).
```

Comment:
Pete and I synced up yesterday to discuss this more. Here is my current draft based on that conversation:

```
+// Writes the syscall fd to a BPF map key created by an active tls call (the SSL_write/SSL_read probes and
+// ensures that set_conn_as_ssl is called. The majority of syscall probes should
+// call this on function entry, however, there are certain probes which are
+// not exclusively used for networking (read/write/readev/writev) and therefore must defer
+// this fd propagation to its ret probe (where it can be validated it is socket related
+// first via sock_event).
+//
+// By ensuring this is called at the appropriate time, we assume the following will be true:
+// 1. BIO native OpenSSL applications will propagate a syscall's fd one or more times for a given tls function
+//    probe depending on if read/write socket calls experience partial reads/writes. We expect that each of
+//    these syscall invocations will be for the same fd until SSL_write/SSL_read returns.
+// 2. Non BIO native cases (nodejs/netty) should never propagate a syscall fd because their SSL_write/SSL_read
+//    probes operate on in memory buffers (io is issued asynchronously). Ensuring that non socket syscalls are
+//    vetted prior to calling this function are crucial to avoid mistaking a file fd (non socket) as a socket
+//    fd in this case.
+// 3. If one or more syscalls occur within SSL_write/SSL_read, we expect the socket fd will always be the same.
+//    This assumption is a requirement for this method of tls tracing and if a fd mismatch were to occur, this
+//    function will detect the scenario and notifies stirling user space.
```

This explanation ended up longer than I expected and I'm starting to think it much be too detailed for this specific function. I think the first paragraph has important context added that was missing, but the enumerated cases in the second section feels slightly out of place to me.

@etep what are your thoughts?

Reply:
Thanks! I ended up trimming the second paragraph. As I mentioned yesterday, I want to update the design doc with all of these findings and I think that will be an easier place to explain the second paragraph (with diagrams and links to our histogram testing).